File: Packets/ChatSend.vb
Public Class ChatSendPacket
    Inherits Packet

    Public Overrides Property OPCode As Byte = &H4
    Public Property Length As Byte = 0
    Public Property World As String = ""
    Public Property ChatChannel As ChatChannel
    Public Property ID As UInt32 = 9001
    Public Property Name As String = ""
    Public Property Text As String = ""

    Sub New(ByVal _World As String, ByVal Channel As ChatChannel, ByVal PlayerName As String, ByVal Chat As String)
        MyBase.New()
        World = _World
        ChatChannel = Channel
        Name = PlayerName
        Text = Chat
    End Sub
    Public Overrides Function GetByteArray() As Byte()
        'Length = (6 + Name.Length + 1 + Text.Length) * 2
        'TextLength = Text.Length
        'NameLength = Name.Length
        'Dim aarray As Byte()
        'Dim barray As Byte() = System.Text.Encoding.UTF8.GetBytes(Text)
        'Dim carray As Byte() = System.Text.Encoding.UTF8.GetBytes(Name)
        'Dim darray As Byte() = {OPCode, Length, 1, 0, 0, 0, 0, 1, NameLength}
        'aarray = Misc.Combine({darray, carray})
        'Dim earray = {TextLength}
        'Return Misc.Combine({aarray, earray, barray})
        Dim p As New PacketBuilder()
        p.Write(CByte(ChatChannel))
        p.Write(World)
        p.Write(ID)
        p.Write(Name)
        p.Write(Text)
        Payload = p.GetBytes
        Return Package(False, True)
    End Function
End Class

File: Packets/ChatSend.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
public class ChatSendPacket : Packet
{

	public override byte OPCode { get; set; }
	public byte Length { get; set; }
	public string World { get; set; }
	public ChatChannel Channel { get; set; }
	public UInt32 ID { get; set; }
	public string Name { get; set; }
	public string Text { get; set; }

	public ChatSendPacket(string _World, ChatChannel _Channel, string PlayerName, string Chat) : base()
	{
		World = _World;
		Channel = _Channel;
		Name = PlayerName;
		Text = Chat;
	}
	public override byte[] GetByteArray()
	{
		PacketBuilder p = new PacketBuilder();
		p.Write(Convert.ToByte(Channel));
		p.Write(World);
		p.Write(ID);
		p.Write(Name);
		p.Write(Text);
		Payload = p.GetBytes();
		return Package();
	}
}
